micropython/usb/usb-device-mouse/usb/device/mouse.py

```diff
@@ -0,0 +1,118 @@
+# MicroPython USB Mouse module
+#
+# MIT license; Copyright (c) 2023-2024 Angus Gratton
+from micropython import const
+import struct
+
+from usb.device.hid import HIDInterface
+
+_INTERFACE_PROTOCOL_MOUSE = const(0x02)
+
+
+class MouseInterface(HIDInterface):
+    # A basic three button USB mouse HID interface
+    def __init__(self, interface_str="MicroPython Mouse"):
+        super().__init__(
+            _MOUSE_REPORT_DESC,
+            protocol=_INTERFACE_PROTOCOL_MOUSE,
+            interface_str=interface_str,
+        )
+        self._l = False  # Left button
+        self._m = False  # Middle button
+        self._r = False  # Right button
+
+    def send_report(self, dx=0, dy=0):
+        b = 0
+        if self._l:
+            b |= 1 << 0
+        if self._r:
+            b |= 1 << 1
+        if self._m:
+            b |= 1 << 2
+        # Note: This allocates the bytes object 'report' each time a report is
+        # sent.
+        #
+        # However, at the moment the HIDInterface class doesn't keep track of each
+        # transfer after it's submitted. So reusing a bytearray() creates a risk
+        # of a race condition if a new report transfer is submitted using the
+        # same buffer, before the previous one has completed.
+        report = struct.pack("Bbb", b, dx, dy)
+
+        super().send_report(report)
+
+    def click_left(self, down=True):
+        self._l = down
+        self.send_report()
+
+    def click_middle(self, down=True):
+        self._m = down
+        self.send_report()
+
+    def click_right(self, down=True):
+        self._r = down
+        self.send_report()
+
+    def move_by(self, dx, dy):
+        if not -127 <= dx <= 127:
+            raise ValueError("dx")
+        if not -127 <= dy <= 127:
+            raise ValueError("dy")
+        self.send_report(dx, dy)
+
+
+# Basic 3-button mouse HID Report Descriptor.
+# This is cribbed from Appendix E.10 of the HID v1.11 document.
+_MOUSE_REPORT_DESC = bytes(
+    [
```

Comment:
I think this report descriptor will compile to far larger code than the descriptor for `usb/device/keyboard.py`. In particular, the (implicit) concatenation of `b"..."`-literals is stored as a constant next to the code, while this variant of constructing the `bytes` object generated a whole bunch of bytecode.

Reply:
You're totally right Felix. This is something I meant to come back to and clearly forgot, appreciate the reminder!

Reply:
Fixed to use same approach as keyboard, thanks again for the reminder!